From bb319b6445a3da35052523a2ac182b47edd2ca6c Mon Sep 17 00:00:00 2001
From: Dunrong Huang <riegamaths@gmail.com>
Date: Wed, 23 Dec 2015 21:42:28 +0800
Subject: [PATCH 13/23] rgw: add a new error code for email conflict

From http://docs.ceph.com/docs/master/radosgw/adminops/ when provided email
address exists, rgw should return EmailExists rather than UserAlreadyExists

Signed-off-by: Dunrong Huang <riegamaths@gmail.com>
(cherry picked from commit c22731698cab7962f3c62934a37d422973ef157f)
Signed-off-by: Dunrong Huang <riegamaths@gmail.com>

Conflicts:
	src/rgw/rgw_common.h
	src/rgw/rgw_user.cc
---
 src/rgw/rgw_common.h      | 1 +
 src/rgw/rgw_http_errors.h | 1 +
 src/rgw/rgw_user.cc       | 7 +++++--
 3 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/rgw/rgw_common.h b/src/rgw/rgw_common.h
index b288d0c..86f5375 100644
--- a/src/rgw/rgw_common.h
+++ b/src/rgw/rgw_common.h
@@ -139,6 +139,7 @@ using ceph::crypto::MD5;
 #define ERR_INVALID_ACCESS_KEY   2028
 #define ERR_MALFORMED_XML        2029
 #define ERR_USER_EXIST           2030
+#define ERR_EMAIL_EXIST          2032
 #define ERR_USER_SUSPENDED       2100
 #define ERR_INTERNAL_ERROR       2200
 
diff --git a/src/rgw/rgw_http_errors.h b/src/rgw/rgw_http_errors.h
index 3a3ef19..a456e72 100644
--- a/src/rgw/rgw_http_errors.h
+++ b/src/rgw/rgw_http_errors.h
@@ -48,6 +48,7 @@ const static struct rgw_http_errors RGW_HTTP_ERRORS[] = {
     { ETIMEDOUT, 408, "RequestTimeout" },
     { EEXIST, 409, "BucketAlreadyExists" },
     { ERR_USER_EXIST, 409, "UserAlreadyExists" },
+    { ERR_EMAIL_EXIST, 409, "EmailExists" },
     { ENOTEMPTY, 409, "BucketNotEmpty" },
     { ERR_PRECONDITION_FAILED, 412, "PreconditionFailed" },
     { ERANGE, 416, "InvalidRange" },
diff --git a/src/rgw/rgw_user.cc b/src/rgw/rgw_user.cc
index bbf2fc7..0769fa1 100644
--- a/src/rgw/rgw_user.cc
+++ b/src/rgw/rgw_user.cc
@@ -1710,12 +1710,15 @@ int RGWUser::execute_add(RGWUserAdminOpState& op_state, std::string *err_msg)
 
     if (op_state.found_by_email) {
       set_err_msg(err_msg, "email: " + user_email + " exists");
+      ret = -ERR_EMAIL_EXIST;
     } else if (op_state.found_by_key) {
       set_err_msg(err_msg, "duplicate key provided");
+      ret = -EEXIST;
     } else {
       set_err_msg(err_msg, "user: " + op_state.user_id + " exists");
+      ret = -EEXIST;
     }
-    return -EEXIST;
+    return ret;
   }
 
   // fail if the user_info has already been populated
@@ -1937,7 +1940,7 @@ int RGWUser::execute_modify(RGWUserAdminOpState& op_state, std::string *err_msg)
       ret = rgw_get_user_info_by_email(store, op_email, duplicate_check);
       if (ret >= 0 && duplicate_check.user_id != user_id) {
         set_err_msg(err_msg, "cannot add duplicate email");
-        return -EEXIST;
+        return -ERR_EMAIL_EXIST;
       }
     }
     user_info.user_email = op_email;
-- 
2.8.0.rc3

